### PR TITLE
Support most Python 2 syntax back to 2.3

### PR DESCRIPTION
**Pull request by thatch**
## Summary

This now parses a good amount of real-world `setup.py` code, and I've validated back to 2.3 (the earliest version I have handy).  It passes most of [python-grammar-changes](https://github.com/thatch/python-grammar-changes) with only a few remaining failures:

```
                                                  23 24 25 26 27 30 31 32 33 34 35 36 37 38
examples/py27-0024-as-keyword.py                  o. o. o  .. .. .. .. .  .. .  .. .. .. .. 
examples/py30-0004-unpacking-args1.py             o. o. o  o. o. .. .. .  .. .  .. .. .. .. 
examples/py30-0013-kwargs-after-star.py           .o .o .  oo oo oo oo o  oo o  oo oo oo oo 
examples/py30-0018-except-as.py                   .o .o .  oo oo oo oo o  oo o  oo oo oo oo 
examples/py30-0023-relative-imports1.py           .o .o o  oo oo oo oo o  oo o  oo oo oo oo 
examples/py30-0023-relative-imports2.py           .o .o o  oo oo oo oo o  oo o  oo oo oo oo 
```

Both `relative-imports` and `except-as` are important.  `unpacking-args` is a feature I haven't seen used since about 2004.  The other two are for additional strictness to reject bad code, but are not necessary for my immediate needs.

Integer validation is reasonably strict at parse time, and handles both the octal changes and `L` suffix change.

```
                                                  23 24 25 26 27 30 31 32 33 34 35 36 37 38
examples/py30-0030-int-literals1.py               oo oo o  oo oo .. .. .  .. .  .. .. .. .. 
examples/py30-0030-int-literals2.py               oo oo o  oo oo .. .. .  .. .  .. .. .. .. 
examples/py30-0030-int-literals3.py               oo oo o  oo oo .. .. .  .. .  .. .. .. .. 
examples/py30-0030-int-literals4.py               .. .. .  oo oo oo oo o  oo o  oo oo oo oo 
examples/py30-0030-int-literals5.py               .. .. .  .. .. .. .. .  .. .  .. oo oo oo 
Legend:
 green header means will test with libcst

 first result is python, second [optional] result is libcst
 o   parses
 .   does not parse
```

New classes have `Py2`-prefixed names, to make it a conscious decision for people so they don't accidentally using them when constructing trees by hand.

## Test Plan

```
$ ../libcst-venv/bin/pyre
 ƛ No type errors found
$ tox -e py36
...
py36 run-test: commands[0] | python -m unittest
...
Ran 1687 tests in 85.662s

OK (skipped=3, expected failures=2)
```

This is relevant to #184

### CODE REVIEW

**Comment by thatch** [comment]
For completeness, here are future import not covered in the `python-grammar-changes` automated tests, which are still to do:

- `unicode_literals` always enabled (needs to store a bit for `evaluated_value` to be correct on unprefixed strings, and probably some allowed escapes changes)
- `absolute_imports` always enabled (which may require changes in the imports codemods)
- `with_statement` always enabled (only affects python 2.5, this was always on in 2.6)

File: libcst/_nodes/expression.py [hunk 1]
```diff
@@ -427,7 +427,7 @@ def _visit_and_replace_children(self, visitor: CSTVisitorT) -> "Integer":
 
     def _validate(self) -> None:
         super(Integer, self)._validate()
-        if not re.fullmatch(INTNUMBER_RE, self.value):
+        if not re.fullmatch(INTNUMBER_RE + "[Ll]?", self.value):
```
Comment (carljm) [inline libcst/_nodes/expression.py]: I guess you decided that slightly lax validation on Py3 was preferable to introducing a separate `Py2Long` node type?

File: libcst/_nodes/statement.py [hunk 2]
```diff
@@ -2409,3 +2503,197 @@ def _codegen_impl(
                 state.add_token("; ")
         elif isinstance(semicolon, Semicolon):
             semicolon._codegen(state)
+
+
+@add_slots
+@dataclass(frozen=True)
+class ExecTarget(CSTNode):
```
Comment (carljm) [inline libcst/_nodes/statement.py]: Why not `Py2ExecTarget`?

File: libcst/_nodes/statement.py [hunk 3]
```diff
@@ -2409,3 +2503,197 @@ def _codegen_impl(
...
+                state.add_token(", ")
+            elif isinstance(comma, Comma):
+                comma._codegen(state)
+
+            target_locals._codegen(state)
+
+
+@add_slots
+@dataclass(frozen=True)
+class Exec(BaseSmallStatement):
```
Comment (carljm) [inline libcst/_nodes/statement.py]: Why not `Py2Exec`?

File: libcst/_parser/conversions/expression.py [hunk 4]
```diff
@@ -861,7 +878,7 @@ def convert_atom_basic(
     elif child.type.name == "NUMBER":
         # We must determine what type of number it is since we split node
         # types up this way.
-        if re.fullmatch(INTNUMBER_RE, child.string):
+        if re.fullmatch(INTNUMBER_RE + "[Ll]?", child.string):
```
Comment (carljm) [inline libcst/_parser/conversions/expression.py]: Shouldn't this be version-specific?

File: libcst/_nodes/expression.py (hunk 1 above)
Comment (thatch) [reply libcst/_nodes/expression.py:430]: Yes; another thing that's more lax than ideal here is that `tokenize.Intnumber` allows underscores (invalid on 2.7 and below) and `0o` prefix (invalid on 2.5 and below).  One relatively common practice in older code is `chmod(0177)` type calls, which isn't handled in this PR (it requires edits to the middle of the regex, at which point we should probably just reproduce it here).  I'm far more concerned with getting the default string type right in a subsequent PR than I am about stricter validation on ints.

File: libcst/_nodes/statement.py (hunk 3 above)
Comment (thatch) [reply libcst/_nodes/statement.py]: Done.

File: libcst/_parser/conversions/expression.py (hunk 4 above)
Comment (thatch) [reply libcst/_parser/conversions/expression.py]: I went ahead and made Py2Integer and split this out.

File: libcst/_nodes/statement.py (hunk 2 above)
Comment (thatch) [reply libcst/_nodes/statement.py]: Done.

File: libcst/_nodes/expression.py (hunk 1 above)
Comment (thatch) [reply libcst/_nodes/expression.py]: There was already some special-casing in the tokenizer, I expanded and made this more complete (see new `0030` series tests in the PR message)

Comment (thatch) [reply libcst/_nodes/expression.py]: I called it Py2Integer though, to lump together long and octal py2-only int literals.

**Comment by carljm** [comment]
Did another more careful run-through, this looks pretty good to me. In an ideal world I'd prefer to bump the test coverage up a bit; a number of the new Py2 nodes don't really get validation or codegen exercised at all by the test suite, and overall this brings test coverage down. Seems like your change-examples diff goes a long way to giving confidence in the correctness here, but I wish we could get more of that confidence built in to LibCST's own test suite so that future changes are validated too. But I'm also not that invested in Py2, either, so not going to block on that :)

Thanks for all the hard work here!

@jimmylai do you want me to hold off on merging so you can review this too, or just go ahead?

**Comment by jimmylai** [comment]
> @jimmylai do you want me to hold off on merging so you can review this too, or just go ahead?

I can review this today. Maybe just wait a bit for me.

**Comment by thatch** [comment]
I will work on increasing coverage, if you want to wait a day...

File: libcst/_nodes/statement.py [hunk 5]
```diff
@@ -2409,3 +2503,197 @@ def _codegen_impl(
...
+    A Python 2 exec statement, such as ``exec "a=1"`` or ``exec "a=1" in
+    globals(), locals()``
+    """
+
+    expr: BaseExpression
+    target: Optional[Py2ExecTarget]
+
+    whitespace_after_exec: SimpleWhitespace = SimpleWhitespace.field(" ")
+
+    semicolon: Union[Semicolon, MaybeSentinel] = MaybeSentinel.DEFAULT
```
Comment (jimmylai) [inline libcst/_nodes/statement.py:2580]: Those attributes miss some detailed Sphinx inline comments like other node type.

File: libcst/_nodes/statement.py [hunk 6]
```diff
@@ -2409,3 +2503,197 @@ def _codegen_impl(
...
+    """
+    """
+
+    target_globals: BaseExpression
+    target_locals: Optional[BaseExpression]
+
+    whitespace_before_in: BaseParenthesizableWhitespace = SimpleWhitespace.field(" ")
+    whitespace_after_in: BaseParenthesizableWhitespace = SimpleWhitespace.field(" ")
+
+    comma: Union[Comma, MaybeSentinel] = MaybeSentinel.DEFAULT
```
Comment (jimmylai) [inline libcst/_nodes/statement.py:2520]: Those attributes miss some detailed Sphinx inline comments like other node type.

**Comment by thatch** [comment]
I'll try to rebase this over the holidays.

File: .gitignore [hunk 7]
```diff
@@ -13,3 +13,4 @@ build/
 .coverage
 .hypothesis/
 .pyre_configuration
+pip-wheel-metadata
```
[the remark this answers is not in the document]
Comment (thatch) [reply .gitignore:16]: I don't remember exactly, but https://github.com/pypa/pip/issues/6213 says "pip wheel ." does it, and that seems reasonable.

**Comment by thatch** [comment]
@jimmylai  I think this is ready for another review.  I fixed some docs and tests, and pyre is passing locally (but I don't entirely believe that).  Is there an easy way to kick off a circleci run?

**Comment by jimmylai** [comment]
> @jimmylai I think this is ready for another review. I fixed some docs and tests, and pyre is passing locally (but I don't entirely believe that). Is there an easy way to kick off a circleci run?

Circle CI shows all tests were passed. It automatically runs whenever you push an update to a PR. Yeah, I'll do another round of review.

**Comment by jimmylai** [comment]
Some functions are not covered by tests. Can you add some tests?
- convert_py2_raise_stmt
- convert_exec_stmt
- Py2Exec. _codegen_impl
- Py2ExecTarget. _validate

https://codecov.io/gh/Instagram/LibCST/pull/275/diff?src=pr&el=tree#diff-bGliY3N0L19faW5pdF9fLnB5

Did you try running this run on our large codebase through Fixit rules and unit tests? That can help us identify some issue from major changes like this PR.
Otherwise, LTGM.

**Pull request closed without merging**